### ISSUE TITLE
integration tests are painfully slow

### ISSUE THREAD

**Issue by BenTheElder** [kubernetes/kubernetes#126202]
This job is often taking well over an hour even with 8 cores and 20Gi assigned in prow. It used to be more like 30m or less.

(because these tests cases are mostly kubectl and apiserver)

A handful of specific tests seem to be large outliers

the CEL tests alone take 11+ minutes
> k8s.io/kubernetes/test/integration/apiserver: cel | 11m13s

/sig api-machinery

cc @jpbetz @cici37 

and OIDC
> k8s.io/kubernetes/test/integration/apiserver: oidc | 11m16s

/sig auth

> k8s.io/kubernetes/test/integration/controlplane: transformation | 14m33s

> k8s.io/kubernetes/test/integration: examples | 7m47s

api machinery again?

> k8s.io/kubernetes/test/integration: job | 14m16s

/sig apps

> k8s.io/kubernetes/test/integration: scheduler_perf | 9m34s

/sig scheduling

Most of the SIG CLI tests take a few seconds

NOTE: I only skimmed a few of the excessive results here, you can see all of the timings if you expand the "174/174 tests passed!" field

https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/126191/pull-kubernetes-integration/1813929954224640000

can we do something about this folks?

/kind bug

**Comment by BenTheElder** [comment]
This sort of thing hurts the contributor experience, we typically would not permit a presubmit that took longer than 1 hour because that creates an excessively long feedback loop.

If you have really slow expensive tests and they must stay that way, they should probably be broken out into another job and we should strongly consider if they need to run on every single PR push versus on merged code. It is not cheap to run this.

**Comment by liggitt** [comment]
> k8s.io/kubernetes/test/integration/controlplane: transformation

cc @enj

**Comment by liggitt** [comment]
> k8s.io/kubernetes/test/integration: scheduler_perf

cc @alculquicondor

**Comment by liggitt** [comment]
> k8s.io/kubernetes/test/integration: job

cc @alculquicondor @mimowo

**Comment by liggitt** [comment]
> k8s.io/kubernetes/test/integration/apiserver: oidc

cc @enj @aramase

**Comment by aojea** [comment]
we need to start splitting between periodics and presubmit, we can tag slows to run only on periodics

**Comment by mimowo** [comment]
A relatively simple optimization for job integration tests might be to reuse the API server across test cases within a test.

**Comment by pohly** [comment]
> It used to be more like 30m or less 

Not for a while. I remember that when I started working on running integration tests with race detection enabled (https://github.com/kubernetes/kubernetes/pull/116980) more than a year ago it was already close to an hour.

> k8s.io/kubernetes/test/integration: scheduler_perf | 9m34s

I was involved with  `scheduler_perf` running as integration test. I added that because it provided some valuable additional coverage of more complex scheduling scenarios. The runtime can be reduced by picking different configurations in performance_config.yaml. What would be an acceptable maximum duration?

> we need to start splitting between periodics and presubmit, we can tag slows to run only on periodics

How? I think we currently have no way to tell `make test-integration` what it should and shouldn't run.

> A relatively simple optimization for job integration tests might be to reuse the API server across test cases within a test.

That's already done.

**Comment by pohly** [comment]
> It was for many many years though, and we should fix that.

I don't disagree. I just thought that "below 1 hour" was the guideline and it was hitting that, at least when I looked.

> `if testing.Short() {`

Neat, I didn't know that one. We don't have any periodic integration test job, do we we?

> performance tests are somewhat less meaningful due to noisy neighbor issues.

scheduler_perf runs in "benchmarking" mode in some other job. During integration testing it runs different configurations of the same scenarios as unit tests.

**Comment by pohly** [comment]
> We don't have any periodic integration test job, do we we?

We do: ci-kubernetes-integration-master

**Comment by alculquicondor** [comment]
> I was involved with `scheduler_perf` running as integration test. I added that because it provided some valuable additional coverage of more complex scheduling scenarios. The runtime can be reduced by picking different configurations in performance_config.yaml. What would be an acceptable maximum duration?

Yeah, we could mark less tests as "fast" so that they don't run as part of the integration job. Should we cap it at 5m? 2m?

**Comment by mimowo** [comment]
> > A relatively simple optimization for job integration tests might be to reuse the API server across test cases within a test.
> 
> That's already done.

Is it? In some tests (see [example](https://github.com/kubernetes/kubernetes/blob/dda657b5982e8f9102b8df5931344262b0793163/test/integration/job/job_test.go#L431-L432)) we call `setup` inside `t.Run` which IIUC spawns the API server inside, see [here](https://github.com/kubernetes/kubernetes/blob/dda657b5982e8f9102b8df5931344262b0793163/test/integration/job/job_test.go#L4232). Though, I'm not sure what is the actual benefit, probably will not change the overall picture much.

**Comment by BenTheElder** [comment]
> Yeah, we could mark less tests as "fast" so that they don't run as part of the integration job. Should we cap it at 5m? 2m?

Let's start at 5m and go from there.

**Comment by alculquicondor** [comment]
/assign @macsko
(for the scheduler_perf side)

**Comment by pohly** [comment]
> >    That's already done.
>
> Is it?

Perhaps not everywhere, but definitely in several tests.

**Comment by macsko** [comment]
scheduler_perf pov:

We have two ways of running the tests:
- benchmark - creating apiserver, etcd and other components per each test case separately ([ref](https://github.com/kubernetes/kubernetes/blob/01eb9f47541156a38e4b28759977c06f07f08012/test/integration/scheduler_perf/scheduler_perf.go#L835-L846))
- integration - creating etcd **once** at the beginning of the tests ([ref](https://github.com/kubernetes/kubernetes/blob/01eb9f47541156a38e4b28759977c06f07f08012/test/integration/scheduler_perf/scheduler_test.go#L42)) and create apiserver for each unique config ([ref](https://github.com/kubernetes/kubernetes/blob/01eb9f47541156a38e4b28759977c06f07f08012/test/integration/scheduler_perf/scheduler_test.go#L80))

I found that running the benchmark way as integration tests with label set to `integration-test`, so running the same test cases, reduces the testing time from 530s to 230s in my case. 

I think the difference might come from cleaning up etcd after each test case which can be time consuming ([ref](https://github.com/kubernetes/kubernetes/blob/01eb9f47541156a38e4b28759977c06f07f08012/test/integration/scheduler_perf/scheduler_perf.go#L1031-L1034)). 

We could unify the tests structure to benchmark way.

**Comment by pohly** [comment]
@macsko: this issue isn't about the "benchmark" mode. That runs in a different job. Restarting etcd and the apiserver during benchmarking is intentional because it reduces the risk that one scenario influences another.

**Comment by pohly** [comment]
> We could unify the tests structure to benchmark way.

That would make it run slower - the opposite of what we want.

**Comment by alculquicondor** [comment]
> I found that running the benchmark way as integration tests with label set to integration-test, so running the same test cases, reduces the testing time from 530s to 230s in my case.

@pohly It looks like, experimentally, just recreating the etcd is faster than removing the objects from the one etcd instance.

**Comment by pohly** [comment]
Ah, okay. That could be true because these tests have a lot of objects to delete, which is done one-by-one instead using deletecollection.

**Comment by kannon92** [comment]
Has there been a consideration to only run these tests on certain changes?

We could separate the long lasting jobs into their own presubmit that runs only when certain changes are made.

And we can leave the full integration job as is for a periodic (if there is a periodic).


For job, I imagine we will continue to add features. In that case, I wonder if having a presubmit that runs on only certain changes could also be an idea to pursue?

I think the work being done is great but I worry that we will end up coming back to this a year later when more features are added.

**Comment by BenTheElder** [comment]
We always have to revisit this for the presubmits, because the imperative is to add more and there isn't readily available tooling to block individually slow tests.

We have periodically revisited this for unit and e2e and verify as well and they all have to be kept in check and we have to balance the value of the tests against reasonable presubmit.

> Has there been a consideration to only run these tests on certain changes?

This is difficult to do reliably. If for example you only run these tests when their own directory changes, that doesn't capture transitive dependencies, go version, etc.

previously sig testing invested in bazel for this purpose, but well that didn't turn out ideally either.

for now we run the slow tests after merge (see above)

**Comment by BenTheElder** [comment]
https://github.com/kubernetes/kubernetes/pull/126305 made a big improvement on the CEL tests, and https://github.com/kubernetes/kubernetes/pull/126451 also got a nice improvement on apiserver/oidc tests.

The job is still really long though, thanks everyone working on these tests.

**Comment by pohly** [comment]
https://github.com/kubernetes/kubernetes/pull/126352 cut down the time for scheduler_perf from >11 min to a bit more than 3 minutes.

**Comment by BenTheElder** [comment]
These are still taking around ~1h10m even on a green run. https://prow.k8s.io/?job=pull-kubernetes-integration

186 test cases: 1h12m:
https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/128786/pull-kubernetes-integration/1867395739894681600

907 test cases: + compiling enough binaries to run an actual cluster and bring it up/down 30m:
https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/122016/pull-kubernetes-e2e-kind/1867418178733215744

Integration testing is a good idea, but this isn't working effectively.

We're basically not using the 8 cores / 20 Gi allocated to this job:
https://monitoring-eks.prow.k8s.io/d/96Q8oOOZk/builds?var-org=kubernetes&var-repo=kubernetes&var-job=pull-kubernetes-integration&orgId=1&refresh=30s

We probably need to parallelize these more generally.

**Comment by pohly** [comment]
I was wondering about that in the context of https://github.com/kubernetes/community/blob/master/contributors/devel/sig-testing/testing-strategy.md:

> E2E Tests: ... These are the most expensive and prone to flakiness.

My experience is that E2E tests are typically easier and cheaper to write, run and debug. Just my two cents... :shrug: 

Regarding existing integration testing: improving parallelization might help. I'm just not sure how without some experimentation. We might also want to reconsider how many different packages with integration tests we have. I remember that at some point @aojea split up some packages to minimize the effect of resource leaks. Perhaps we are in a better shape now and can reduce the number of different packages again that `go test` needs to test?

**Comment by aojea** [comment]
> My experience is that E2E tests are typically easier and cheaper to write, run and debug. Just my two cents... 🤷

you can not do table tests, are slower, are not hermetic ... there are no better or worse here, using them as replacement of the other is a bad decision, each of them fulfill a different purpose ... I have plenty of examples of projects with flaky CIs because they use e2e only that are spending their top engineers chasing flakes because the other teams are not able to debug them ...

**Comment by pohly** [comment]
> you can not do table tests

Sure you can. Just register different `ginkgo.It`s with different parameters, or use [ginkgo.DescribeTable](https://pkg.go.dev/github.com/onsi/ginkgo/v2#DescribeTable).

> are slower

That's not backed by the observation that the E2E kind job runs more tests/min compared to the integration job *and* stands up a cluster for doing it. Of course it could be an apple/orange comparison.

It seems we've had quite different experiences. I don't doubt that. But it makes me wary of pushing people towards writing integration tests when an E2E test would also work.

**Comment by liggitt** [comment]
Running an individual e2e test from zero is definitely slower than an individual integration test ... setup for an integration test is O(seconds). If a scenario can be tested with an integration test, I would *always* recommend that over an e2e.

**Comment by BenTheElder** [comment]
> That's not backed by the observation that the E2E kind job runs more tests/min compared to the integration job and stands up a cluster for doing it. Of course it could be an apple/orange comparison.

the comparison here is the effective performance of these, which appear to be overly setualized. E2E has more overhead and I'm not advocating that we switch to e2e, I'm advocating that we fix integration test throughput.


> Running an individual e2e test from zero is definitely slower than an individual integration test ... setup for an integration test is O(seconds). If a scenario can be tested with an integration test, I would always recommend that over an e2e.

Yeah, the problem is we don't seem to be running the full set effectively. Like based on utilization it appears you could divide the tests into 4 groups and run those on the same machine in different containers and still have resources to spare.

**Comment by BenTheElder** [comment]
e2e has LOTs of slow tests, but we also run with better concurrency. That's a problem in how we orchestrate the execution. We should still try to speed up excessively slow integration tests but I think to get better total runtime we have to pursue more concurrency.

If we don't fix that then at some point adding more integration tests becomes a problem even if they're otherwise preferential and often individually very fast (many are a few seconds but some are still on the order of 10m), because executing all of them together is going to take too long (I would argue >1h is already too long and painful for contributors and a merge bottleneck, though GCE e2e is at nearly 1h currently)

**Comment by mimowo** [comment]
Maybe a relatively easy path to make integration tests run in parallel is to split them by components -kube-scheduler, apiserver and controller manager (maybe also kubectl?) - per CI job.

It is not a deeply thought idea, and have downsides like lower parallelism than if we run in parallel all of them in one suite, but maybe execution of the idea could be simpler, wdyt?

**Comment by BenTheElder** [comment]
https://github.com/kubernetes/kubernetes/issues/130755

we've been setting GOMAXPROCS=4, when the job has `cpu: 8`

fix: https://github.com/kubernetes/kubernetes/pull/130758

**Comment by BenTheElder** [comment]
ran in 35m :-)) https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/130758/pull-kubernetes-integration/1899925030251270144

**Comment by BenTheElder** [comment]
See also: https://github.com/kubernetes/test-infra/pull/34387

**Comment by BenTheElder** [comment]
<30m would still be a better sweet spot, but we're _way_ closer now, and finally competitive with e2e

**Comment by pacoxu** [comment]
> See also: [kubernetes/test-infra#34387](https://github.com/kubernetes/test-infra/pull/34387)

With split of test-cmd and integration:
- [pull-kubernetes-cmd-canary](https://prow.k8s.io/job-history/gs/kubernetes-ci-logs/pr-logs/directory/pull-kubernetes-cmd-canary): ~13-15 minutes
- [pull-kubernetes-integration-canary](https://prow.k8s.io/job-history/gs/kubernetes-ci-logs/pr-logs/directory/pull-kubernetes-integration-canary): ~57 minutes
- [pull-kubernetes-integration](https://prow.k8s.io/job-history/gs/kubernetes-ci-logs/pr-logs/directory/pull-kubernetes-integration): ~1h6m-1h10m

**Comment by pohly** [comment]
> we've been setting GOMAXPROCS=4, when the job has cpu: 8
>
> fix: https://github.com/kubernetes/kubernetes/pull/130758

The pull-kubernetes-integration job definition sets `KUBE_INTEGRATION_TEST_MAX_CONCURRENCY=8`. Should we drop that and rely only on the `ncpu` tool? It also detects 8 cores, so the fixed `KUBE_INTEGRATION_TEST_MAX_CONCURRENCY` is redundant and just another thing which needs to be kept in sync.

**Comment by pohly** [comment]
> <30m would still be a better sweet spot, but we're way closer now, and finally competitive with e2e

Using GOMAXPROCS=16 instead of the actual 8 cores brings it down to 25 minutes, without test failures. Experiment in https://github.com/kubernetes/kubernetes/pull/130770.

**Comment by BenTheElder** [comment]
Unfortunately we've had to revert the KUBE_INTEGRATION_TEST_MAX_CONCURRENCY (which overrides GOMAXPROCS) change for now https://github.com/kubernetes/kubernetes/pull/130821

**Comment by BenTheElder** [comment]
IMHO we should attempt to revisit this early in 1.24, after we drain the code-thaw queue.

**Comment by enj** [comment]
> > k8s.io/kubernetes/test/integration/controlplane: transformation
> 
> cc [@enj](https://github.com/enj)

Current timing of transformation tests: https://gist.github.com/enj/821ed4dff5b901a7ebb49c838315f0a4

There is clearly room for improvement here.

**Comment by BenTheElder** [comment]
cc @ylink-lfs

**Comment by BenTheElder** [comment]
Recent run takes >1h still:
https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/131458/pull-kubernetes-integration/1942364265927151616

If we could figure out the flakiness issues and re-enable GOMAXPROCS auto-detection instead of hardcoding, we could probably more than halve the duration of this job again.

**Comment by ylink-lfs** [comment]
Sync a benchmarking result here: 
https://github.com/kubernetes/kubernetes/pull/130758#issuecomment-3060810578

**Comment by BenTheElder** [comment]
https://github.com/kubernetes/test-infra/pull/35416 moved these jobs to faster hardware, which will give us some nice time improvements, between that and splitting into the two jobs we are finally down under 1h. Thanks @upodroid !

There is still room for improvement, we should revisit the hardcoded maxprocs on these machines.

https://github.com/kubernetes/kubernetes/pull/133759

**Comment by BenTheElder** [comment]
We still need more test runs, but I think we can just re-apply https://github.com/kubernetes/kubernetes/pull/130758, we have new machines with faster cores but also newer disk.

So far with that re-applied, we're getting <= 30m runs in https://github.com/kubernetes/kubernetes/pull/133759, no failures yet in 3 attempts, a fourth is running now.

Which is down from the ~47m run we got just from switching the machines https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/131639/pull-kubernetes-integration-canary/1960979363843280896
